docs/madr/decisions/086-endpoint-to-retrieve-a-resource-by-kri.md

```diff
@@ -0,0 +1,175 @@
+# Add REST Endpoint for KRI-based Resource Lookup
+
+* Status: Accepted
+
+Technical Story: [https://github.com/kumahq/kuma/issues/13882](https://github.com/kumahq/kuma/issues/13882)
+
+## Context and Problem Statement
+
+We want to enable clients (GUI, SDKs, CLI) to fetch any resource using its **Kuma Resource Identifier (KRI)**.
+Currently, clients construct REST paths using mesh and resource name (`/meshes/{mesh}/{resourceType}/{name}`).
+Because of this, it is not possible to simply link between KRI resources: if you have a KRI in one place, you cannot directly navigate to that resource.
+This complicates GUI/frontend development.
+
+A dedicated endpoint for retrieving a resource by its KRI would:
+
+* Provide a canonical API surface for resource lookups.
+* Simplify frontend logic.
+* Ensure KRIs become first-class citizens in the system.
+
+Use cases:
+
+* GUI fetching a resource by clicking a link derived from a KRI.
+* CLI retrieving resources directly by KRI without constructing full paths.
+* SDKs needing stable and consistent access patterns across resource types.
+
+## Decision Drivers
+
+* Ensure stable, predictable API for GUI and SDKs.
+* Keep API design minimal, avoid unnecessary complexity.
+* Maintain performance (avoid costly label lookups).
+* Avoid breaking existing clients; additive change only.
+* Align with OpenAPI/Typescript code generation while balancing developer ergonomics.
+
+## Alternatives Considered
+
+### Backend
+
+#### Option A: Parse KRI and reuse existing store logic (not possible)
+
+**Idea**: Backend parses KRI into `mesh` + `name` and calls existing `store.GetByKey` logic.
+
+* Not possible because the name in KRI is derived using `GetDisplayName` which relies on the `kuma.io/display-name` label.
```

Comment:
If you are _parsing_ a KRI I don't understand why this has anything to do with display-name specifically. 

KRIs have a straightforwards `_` separator and ordered parameters i.e. `kri_<type/shortName>_<mesh>_<zone>_<ns>_<name>_<sectionName>`

Is this a reference to the fact that currently resources are stored using a `<name>-<black-box-hash-containing-at-least-zone-and-possibly-other-things>`? Therefore you can't only use the <name> part of the parsed KRI, you need <name> plus <hash>.

I'm just guessing here as to what the problem could be. Let me know!

Reply:
> Is this a reference to the fact that currently resources are stored using a <name>-<black-box-hash-containing-at-least-zone-and-possibly-other-things>? Therefore you can't only use the part of the parsed KRI, you need <name> plus <hash>.

More or less yes but apparently Ilya has some way of working around this. I'll sync up with him about this today.

Reply:
This can be implemented as one function. Takes a KRI and returns a [ResourceKey](https://github.com/kumahq/kuma/blob/9e9ad8aadf73f240763c30174cfed7ea7ef416eb/pkg/core/resources/model/resource.go#L42-L45).

1. Figure out if resource originated on this CP or not. This can be done via a function like "IsLocallyOriginated"
2. If it's locally originated you just need to adjust for the k8s store because on k8s we have CoreName (which is `${name}.${namespace}`). If it's universal that's not the case.
3. If it's not locally originated you need to compute the hash from the data from KRI. https://github.com/kumahq/kuma/blob/c989d3d842850aa468248e76298b9729af217a3b/pkg/kds/context/context.go#L232 . Here you also need to adjust for CoreName.